### PR TITLE
WIP: Introduce a Reason/Bucklescript loader

### PR DESCRIPTION
**Pull request by dmsnell**
Shhh! this is an experiment…

**What?**

Reason/ReasonML is an alternative syntax for the OCaml language. BuckleScript is a compile target for the OCaml compiler. Thus, this is all just OCaml compiled to JavaScript. Reason-React offers familiar JSX syntax within the Reason modules to make React components in a very similar manner as we make them in JavaScript currently.

**Why?**

JavaScript isn't getting us there. We keep adding layers on top of layers of different tools and linters and libraries to make for the shortcomings of the language in what we want to do and in order to try and contain its sprawl. This PR is an attempt to explore the idea of stepping back and seeing if there's another good way forward waiting under our nose.

 - Less fatigue. This big jump is qualitatively different than other jumps. OCaml has been stable for a very long time. It can be measured not by what it _adds_, but by what it _takes away_, including but not limited to the need for immutable libraries, lodash, babel, eslint, and prettier.

 - Speed. Our builds are routinely taking 10s - 30s for me (even on incremental builds). The claim is that Reason is at least ten times faster than TypeScript when compiling, yet it offers _more safety_.

 - Gradual introduction. Partly on account of how fast it is, we can gradually pull in tiny pieces of Reason code into the project without significant alteration of the build process. Further, since the compiler is supposedly so much faster than our Babel+eslint pipeline it probably makes the build _faster_ by introducing this here.

 - IDE integration/development workflow. OCaml has an interactive REPL to play in and IDE support for autocomplete and error reporting is great.

 - Friendly static typing. We have _tens of thousands_ of runtime type errors every day with Calypso. These bugs can be 100% automatically detected yet we are trying to manually catch them. OCaml's type system (the inspiration for Haskell, Elm, PureScript, Swift, etc…) is incredibly friendly and unlike anything experienced in C/C++/Java. It _helps_ much and _requires_ little. Algebraic data types are amazing.

 - Balance between purely-functional and imperative programming paradigms.

 - Readable output. JavaScript produced by Bucklescript is considerably more readable than what Babel produces. Some at Facebook are apparently checking in the generated JavaScript as if it were hand-written.

 - Embraced by Facebook. 20% of the Messenger app is supposedly written in Reason already. It has JSX and native React bindings.

 - Completeness. OCaml is a robust language with a mature compiler and optimization pipeline. JavaScript semantics make writing good compilers and optimizations incredibly difficult. Instead of trying to "make it work" we just let it do what it's been doing well for a long time.

**Drawbacks**

 - React components are best understood in the ReasonML code itself because we are typically less familiar with the non-JSX style of component definition. That is, the generated JS of React components is unfamiliar.

 - Data types clash here especially with the example in the reducer. If we want an action type data type with the list of available types then we have to copy the entire list _plus_ ReasonML is treating them as an integer enum instead of a string enum. This is ideal of course, but since we have to coexist with the existing state it leads to tricky situations where we have to convert data types.

**Where?**

My initial pursuit a couple months ago was prompted by work I was doing in our state tree and I felt that a compiled language could produce a much better reducer for our app than hand-written JavaScript. Trying out Reason has led me to believe that it too is a viable way to write React components from the large to the small.

Interop requires a few additional steps but it's not difficult. Places that are JS-heavy vs. React-heavy could be great candidates for Reason code due to the optimizations in the compile process.

_Remember:_ Reason/OCaml/Bucklescript is performing considerable safety checking automatically for us so that we can have confidence that our code won't throw an exception when it runs, without _any_ runtime checks.

**Testing**

<del>Open https://calypso.live/comments?branch=explore/introduce-reason</del> <ins>the live branch doesn't seem to be working, so build locally and go to `/comments`</ins>

Useful links:
 - [Reason](https://facebook.github.io/reason/index.html)
 - [Reason-React](https://reasonml.github.io/reason-react/#reason-react)
 - [Reductive (Redux lib)](https://github.com/reasonml/reductive)
 - [Converting from JS](http://facebook.github.io/reason/jsConversion.html)

**Actual generated JS**
```js
// Generated by BUCKLESCRIPT VERSION 1.8.2, PLEASE EDIT WITH CARE
'use strict';

var Curry       = require("bs-platform/lib/js/curry.js");
var React       = require("react");
var ReasonReact = require("reason-react/lib/js/src/reasonReact.js");

var component = ReasonReact.statelessComponent("ReasonableDemo");

function make(name, count, onClick, _) {
  var clickIt = function (_, _$1) {
    return onClick();
  };
  var newrecord = component.slice();
  newrecord[/* render */9] = (function (self) {
      return React.createElement("div", undefined, React.createElement("label", undefined, name), React.createElement("button", {
                      className: "button is-compact",
                      type: "button",
                      onClick: Curry._1(self[/* update */2], clickIt)
                    }, " -> " + count));
    });
  return newrecord;
}

var comp = ReasonReact.wrapReasonForJs(component, (function (jsProps) {
        return make(jsProps.name, jsProps.count, jsProps.onClick, /* array */[]);
      }));

function counter(state, action) {
  var prev = state !== undefined ? state : 0;
  var match = action.type;
  switch (match) {
    case 0 : 
        return prev + 1 | 0;
    case 1 : 
        return prev - 1 | 0;
    case 2 : 
        return prev + action.amount | 0;
    
  }
}

var counter_Increment = /* Counter_Increment */0;

var counter_Decrement = /* Counter_Decrement */1;

var counter_Add = /* Counter_Add */2;

exports.component         = component;
exports.counter_Increment = counter_Increment;
exports.counter_Decrement = counter_Decrement;
exports.counter_Add       = counter_Add;
exports.make              = make;
exports.comp              = comp;
exports.counter           = counter;
/* component Not a pure module */
```

### CODE REVIEW

File: bsconfig.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+{
+  "name": "wp-calypso",
+  "sources": [
+    "client/bs"
+  ],
+  "bs-dependencies": [
+    "reason-react"
+  ],
+  "reason": {
+    "react-jsx": true
```
Comment (chenglou) [inline bsconfig.json]: Try `"react-jsx": 2` here. I forgot to update that in bs-loader just now.

Comment (dmsnell) [inline bsconfig.json]: was going to ask after testing. I found conflicting docs suggesting this

File: bsconfig.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
+/* bsconfig.json */
+{
+  "name": "wp-calypso",
+  "sources": [
+    "client/bs"
```
Comment (chenglou) [inline bsconfig.json]: I'm not sure this works as intended? Schema at http://bloomberg.github.io/bucklescript/docson/#build-schema.json

`[{dir: "client", subdirs: "bs"}]`

File: bsconfig.json (hunk 1 above)
Comment (chenglou) [inline bsconfig.json]: Conflicting docs? Where?

Comment (dmsnell) [inline bsconfig.json]: https://github.com/rrdelaney/bs-loader

but you can see that @rrdelaney updated that in https://github.com/rrdelaney/bs-loader/commit/9134552830723ff313a988d32e70d15db5fe1bd8 recently

File: package.json [hunk 3]
```diff
@@ -249,6 +251,8 @@
     "react-hot-loader": "1.3.0",
     "react-test-env": "0.2.0",
     "readline-sync": "1.4.5",
+    "reason-js": "0.3.0",
```
Comment (rrdelaney) [inline package.json]: `reason-js` can be removed 🙂 The `bs-loader` docs were outdated and @chenglou recently updated them to match the current API and standards

**Comment by dmsnell** [comment]
Thanks again for the tips @chenglou and @rrdelaney - updated the PR in b0f382d to take those into account.

File: webpack.config.js [hunk 4]
```diff
@@ -65,11 +65,15 @@ const webpackConfig = {
 			{
 				include: /node_modules[\/\\]tinymce/,
 				loader: 'imports?this=>window',
+			},
+			{
+				test: /\.(ml|re)$/,
```
Comment (yawaramin) [inline webpack.config.js]: You'll probably also want to add the `.mli` and `.rei` extensions to load any module interface files. Those are like C header files and are super useful in OCaml.

Comment (dmsnell) [inline webpack.config.js]: @yawaramin would we add them here so they go through the loader?

Comment (yawaramin) [inline webpack.config.js]: Actually, I just checked with @rrdelaney and it looks like you don't need to explicitly list the interface files--bs-loader automatically respects them.

Comment (rrdelaney) [inline webpack.config.js]: Yep! BuckleScript will compile using the `.rei` or `.mli` file, but when referring to the file with `bs-loader` use the implementation

File: client/components/reasonable-demo/reasonableDemo.re [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  | Counter_Decrement
+  | Counter_Add
+  [@@bs.deriving {accessors: accessors}];
+
+let make ::name ::count ::onClick _children => {
+  let text = ReasonReact.stringToElement;
+  let clickIt _event _state _self => onClick () [@bs];
+  {
+    ...component,
+    render: fun () self =>
```
Comment (chenglou) [inline client/components/reasonable-demo/reasonableDemo.re]: Small note: we've released a new update that turns this into `render: fun self =>`. No need for the `()` anymore. If you change `reason-react` to `^0.2.3` then it'll Just Work.

Comment (dmsnell) [inline client/components/reasonable-demo/reasonableDemo.re]: updated! thanks!

**Comment by dmsnell** [comment]
Deferring for the future 🚀

**Pull request closed without merging**